ports/libmlservo/src/lib.rs

```diff
@@ -177,15 +212,46 @@ pub unsafe extern "C" fn heartbeat_servo(servo: *mut ServoInstance) {
                 EmbedderMsg::NewFavicon(..) |
                 EmbedderMsg::HeadParsed |
                 EmbedderMsg::SetFullscreenState(..) |
-                EmbedderMsg::ShowIME(..) |
-                EmbedderMsg::HideIME |
                 EmbedderMsg::Shutdown |
                 EmbedderMsg::Panic(..) => {},
             }
         }
     }
 }
 
+#[no_mangle]
+pub unsafe extern "C" fn keyboard_servo(
+    servo: *mut ServoInstance,
+    key_code: char,
+    key_type: MLKeyType,
+) {
+    if let Some(servo) = servo.as_mut() {
+        let key = match key_type {
+            MLKeyType::kCharacter => Key::Character([key_code].iter().collect()),
+            MLKeyType::kBackspace => Key::Backspace,
+            MLKeyType::kEnter => Key::Enter,
+            _ => return,
+        };
+
+        let key_down = KeyboardEvent {
+            state: KeyState::Down,
+            key: key,
+            ..KeyboardEvent::default()
+        };
+
+        let key_up = KeyboardEvent {
+            state: KeyState::Up,
+            ..key_down.clone()
+        };
+
+        // TODO: can the ML1 generate separate press and release events?
```

Comment:
https://github.com/webmixedreality/exokit/blob/master/deps/exokit-bindings/magicleap/src/magicleap.cc#L1451-L1473 suggests yes.

Reply:
That's using their own keyboard event impl, not the lumin runtime one. https://github.com/webmixedreality/exokit/blob/106bfbb4829057906067e6a88e4450df3e1895c2/deps/exokit-bindings/magicleap/include/magicleap.h#L86 uses https://github.com/webmixedreality/exokit/blob/106bfbb4829057906067e6a88e4450df3e1895c2/deps/exokit-bindings/magicleap/include/magicleap.h#L69 which has separate events for up and down.

The matching lumin events only have a pressed event: https://creator.magicleap.com/learn/reference/public/v0.16.0/LuminRT/namespacelumin_1_1____lr1__0_1_1ui_1_1_keyboard_event.html